Add melcloud standard horizontal vane modes

antlarr opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This commit allows users that have HVACs devices managed with melcloud to use the standard climate card to set the horizontal vane mode.

This just implements the standard methods added recently for horizontal vane support (as explained in https://developers.home-assistant.io/blog/2024/12/03/climate-horizontal-swing/ which was very helpful in knowing of this new support) in melcloud.



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

I tested this in my own melcloud devices.

In the future it would be interesting to remove the custom vane support in the melcloud integration and only use the standard services/methods but this would break existing automations so IMO it should be deprecated and cleaned up in a separate PR to be merged in a future version. In the meantime, it doesn't hurt to have both ways to access the same states. If someone points me to some documentation on how to deprecate services, I'll be glad to submit a PR too for that.




- This PR fixes or closes issue: fixes #123938
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

antlarr committed:
Add support for standard horizontal vane modes

Since homeassistant recently added standard way to implement
horizontal vane support in climate entities, this commit
adds support for those in melcloud devices.

I own a device that is managed with the melcloud integration
where I tested this.

In the future it would be interesting to remove the custom
vane support in the melcloud integration and only use the
standard services/methods but this would break existing
automations so IMO it should be deprecated and cleaned up
in a future version. In the meantime, it doesn't hurt to
have both ways to access the same states.

antlarr committed:
Fix parameter name to use swing_horizontal_mode

Fix parameter name to use swing_horizontal_mode in async_set_swing_horizontal_mode

joostlek commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -162,6 +162,7 @@ def extra_state_attributes(self) -> dict[str, Any] | None:
                     ATTR_VANE_HORIZONTAL_POSITIONS: self._device.vane_horizontal_positions,
                 }
             )
+            self._attr_supported_features |= ClimateEntityFeature.SWING_HORIZONTAL_MODE
```
We should not set this during `extra_state_attributes` as that is a side effect. If for example extra_state_attributes wouldn't be called for what reason, the supported feature is not added, and that is bad.

antlarr replied:
Before submitting this PR I tried checking for self._device.vane_horizontal (to check if the device supports horizontal vanes) in __init__ but that didn't work since the device isn't initialized until the device information is obtained from the cloud.

So I tried reimplementing the MelCloudClimate.async_update in AtaDeviceClimate and check for horizontal vane support when the MelCloudClimate.async_update returns. I tested that works fine here too, so I submitted that change in a new commit.

antlarr replied:
I've marked this as resolved since I submitted a fix for the issue you commented. If you think it's still wrong, please reopen it.

antlarr committed:
Use async_added_to_hass to set horizontal swing as supported

Reimplement async_added_to_hass in AtaDeviceClimate (from MelCloudClimate)
so that when the device information is obtained from the cloud, we
check if the device has horizontal vanes to mark SWING_HORIZONTAL_MODE
as a supported feature.

Note that we can't just check for self._device.vane_horizontal in
__init__ since vane_horizontal is only set (or not) after the device
data is fetched from the cloud.

antlarr commented:
@joostlek, is there anything missing here that I need to fix? afaik, it should be fine now.

joostlek commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -150,6 +150,14 @@ def __init__(self, device: MelCloudDevice, ata_device: AtaDevice) -> None:
         self._attr_unique_id = f"{self.api.device.serial}-{self.api.device.mac}"
         self._attr_device_info = self.api.device_info
 
+    async def async_update(self) -> None:
+        """Update state from MELCloud."""
+        await super().async_update()
+
+        # We can only check for vane_horizontal once we fetch the device data from the cloud
+        if self._device.vane_horizontal:
+            self._attr_supported_features |= ClimateEntityFeature.SWING_HORIZONTAL_MODE
```
So if the supported feature can change, I would recommend having a `supported_features` property. If it can't change I would recommend having it in `async_added_to_hass`

antlarr replied:
ah, thanks, I didn't know about `async_added_to_hass`. I don't think the feature can change, so I've modified the commit to use that method instead of `async_update` (and tested that it works fine too)

joostlek commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -274,15 +282,29 @@ def swing_mode(self) -> str | None:
 
+    @property
+    def swing_horizontal_modes(self) -> list[str] | None:
+        """Return a list of available vertical vane positions and modes."""
```
```suggestion
        """Return a list of available horizontal vane positions and modes."""
```

joostlek commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -274,15 +282,29 @@ def swing_mode(self) -> str | None:
         await self.async_set_vane_vertical(swing_mode)
 
+    async def async_set_swing_horizontal_mode(self, swing_horizontal_mode: str) -> None:
+        """Set vertical vane position or mode."""
```
```suggestion
        """Set horizontal vane position or mode."""
```

joostlek commented on homeassistant/components/melcloud/climate.py:
```diff
@@ -274,15 +282,29 @@ def swing_mode(self) -> str | None:
 
+    @property
+    def swing_horizontal_mode(self) -> str | None:
+        """Return vertical vane position or mode."""
```
```suggestion
        """Return horizontal vane position or mode."""
```

joostlek committed:
Apply suggestions from code review

joostlek commented:
> , it doesn't hurt to have both ways to access the same states. If someone points me to some documentation on how to deprecate services, I'll be glad to submit a PR too for that.

A proper way to deprecate this is to create an issue when that service is called. It can tell the user that the service is deprcated and explain what should be used from now on. Deprecation period is 6 months. (and I think it would be nice to deprecate before the summer, since I would assume most automations related to this are run in the summer months, so if we deprecated this in september we would have people who missed this)
